assignments/undo.md

```diff
@@ -30,48 +30,83 @@ Tasks:
+`ColorPicker` you have already implemented. For an example of how this
+can look, look at screenshots 3 and 4.
+
+When you click the FAB to select the stroke thickness, the user should be
+able to decide between three thicknesses (each with their own FAB). For
+an example of how this should look, look at screenshots 5 and 6.
+
+#### Handling Undo/Redo Operations
+Show the undo FAB if and only if there is any command to undo. See the behavior in screenshots 6 and 7 for an example of what happens when
+the undo FAB is clicked. On a long press of the undo button, a list of
+recent commands (up to 10) is displayed. Clicking a command should undo all actions up to and including the selected action.
+
+Show the redo FAB if and only if there is any command to redo. See the behavior in screenshots 7 and 8 for an example of what happens when the
+redo FAB is clicked. On a long press of the redo button, a list of recent commands (up to 10) is displayed. Clicking a command should redo all actions up to and including the selected action.
+
+There should be a limit on how many actions one can undo or redo. You should only allow up to 10 commands to undo and up to 10 commands to redo.
+
+#### FAB States
+ - When there is no command to undo, the undo FAB should not be visible.
+ - When there is no command to redo, the redo FAB should not be visible.
+ - As a color is selected and after the color is selected, the color FAB
+ should update its background to that color.
+ - When a thickness is picked, the thickness FAB should update its icon
+ to indicate the thickness selected.
+
+## Commands
+There are three types of Commands. Each of these commands are extended from the Command abstract class.
+
+- `CommandChangeColor`
+- `CommandChangeThickness`
+- `CommandDrawPath`
+
+## Designing the Canvas State
+Implement the constructor for the CanvasState. A CanvasState stores
+all the commands needed to generate the current canvas.
+
+Then implement `drawCanvas(Canvas canvas)`. This function will apply
+the commands that a CanvasState is comprised of to draw on the canvas.
+
+## Managing Different CanvasStates with a CanvasStateManager
+The CanvasStateManager will handle moving back and forth between `CanvasState`s as the user undoes, redoes, or applies actions. The
+`DrawingView` instance has a `CanvasStateManager.`
+
+`getUndoHistory()` and `getRedoHistory()` returns an array of the descriptors for the commands one can undo and redo respectively.
+
+`applyNewCommand(Command command)` will apply a new Command. This entails
+creating a new CanvasState accounting for the command and updating
+the `currentCanvasState`, `undoStack`, and `redoStack`.
+
+`undo()` and `redo` apply an undo and redo respectively to the `CanvasStateManager`. This entails
+updating the `currentCanvasState`, `undoStack`, and `redoStack`.
+
+## Screenshots of the user experience
 
 ![Screenshot of step 1](undo-img/1.png){:width="200px"}
```

Comment:
I know you didn't add these, but could you change these to use percentage widths like in Doodle and Layout? It's just cleaner

Reply:
I will fix this in its own commit.